racket/collects/racket/bytes.rkt

```diff
@@ -2,9 +2,22 @@
 
 (provide bytes-append* bytes-join)
 
+(define (check-bytes-list* l name)
+  (let loop ((l l) (p null))
+    (cond ((null? (cdr l))
+           (define tail (car l))
+           (or (and (list? tail) (andmap bytes? tail))
+               (raise-argument-error name "(listof bytes?)" tail))
+           (foldl cons tail p))
+          (else
+           (define e (car l))
+           (or (bytes? e)
+               (raise-argument-error name "bytes?" e))
+           (loop (cdr l) (cons e p))))))
+
 (define bytes-append*
-  (case-lambda [(strs) (apply bytes-append strs)] ; optimize common case
-               [(str . strss) (apply bytes-append (apply list* str strss))]))
+  (case-lambda [(strs) (apply bytes-append (check-bytes-list* (list strs) 'bytes-append*))] ; optimize common case
+               [(str . strss) (apply bytes-append (check-bytes-list* (cons str strss) 'bytes-append*))]))
```

Comment:
The common case isn't much of an optimization anymore, so maybe keep just the second one?